app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,274 @@
+
+            //Make sure there are no zeroes in the passphrase
+            //SQLCipher passphrases must not contain any zero byte-values
+            //For more, see "Creating the Passphrase" section here:
+            //https://commonsware.com/Room/pages/chap-passphrase-001.html
+            boolean containsZero = false;
+            for (byte b : result) {
+                if (b == 0) {
+                    containsZero = true;
+                    break;
+                }
+            }
+
+            if (!containsZero) {
+                break;
+            }
+        }
+
+        return result;
+    }
+
+    private static String getTransformationString() {
+        String transformation;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            transformation = String.format("%s/%s/%s", ALGORITHM, BLOCK_MODE, PADDING);
+        }
+        else {
+            transformation = "RSA/ECB/PKCS1Padding";
+        }
+
+        return transformation;
+    }
+
+    //Encryption backed by Android KeyStore
+    private static byte[] encrypt(Context context, byte[] bytes)
+            throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException,
+            IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException,
+            UnrecoverableEntryException, CertificateException, KeyStoreException, IOException, NoSuchProviderException {
+        String transformation = getTransformationString();
+        Cipher cipher = Cipher.getInstance(transformation);
+        cipher.init(Cipher.ENCRYPT_MODE, getKey(context, true));
+        byte[] encrypted = cipher.doFinal(bytes);
+        byte[] iv = cipher.getIV();
+        int ivLength = iv == null ? 0 : iv.length;
+
+        byte[] output = new byte[encrypted.length + ivLength + 3];
+        int writeIndex = 0;
+        output[writeIndex] = (byte) ivLength;
+        writeIndex++;
+        if(ivLength > 0) {
+            System.arraycopy(iv, 0, output, writeIndex, iv.length);
+            writeIndex += iv.length;
+        }
+
+        output[writeIndex] = (byte) (encrypted.length / 256);
+        writeIndex++;
+        output[writeIndex] = (byte) (encrypted.length % 256);
+        writeIndex++;
+        System.arraycopy(encrypted, 0, output, writeIndex, encrypted.length);
+
+        return output;
+    }
+
+    //Decryption backed by Android KeyStore
+    private static byte[] decrypt(Context context, byte[] bytes)
+            throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidAlgorithmParameterException,
+            InvalidKeyException, IllegalBlockSizeException, BadPaddingException,
+            UnrecoverableEntryException, CertificateException, KeyStoreException, IOException, NoSuchProviderException {
+        int readIndex = 0;
+        int ivLength = bytes[readIndex];
+        readIndex++;
+        if(ivLength < 0) {
+            //TODO: Early chance to catch decryption error
```

Comment:
probably should throw a specific exception from here for callers to catch and act on

Reply:
Done